### PR TITLE
fix(es/minifier): Fix detection of direct `eval`

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

The existence of `eval` make the name mangler abort, but partial try logic causes minifier to generate an invalid code.

This is what I get if I disable eval detection.

image

With eval detection, I get 

```js

                     function r(r) {
                        var o = e[r];
                        if (void 0 !== o) return o.exports;
                        var n = e[r] = {
                            exports: {}
                        };
                        return t[r](n, n.exports, r), n.exports;
                    }
```

which is wrong

**Related issue:**

 - https://github.com/vercel/next.js/discussions/30237#discussioncomment-3920682

---

# Investigation

image

## Conclusion

`eval` breaks the name mangler.

### CODE REVIEW

File: crates/swc_ecma_minifier/tests/terser/compress/issue_973/this_binding_collapse_vars/output.mangleOnly.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
-var c = a;
-c();
-var d = a.b;
-d();
-var e = eval;
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/terser/compress/issue_973/this_binding_collapse_vars/output.mangleOnly.js:5]: This is an indirect eval, so it cannot use `c` nor `b`

File: crates/swc_ecma_minifier/tests/terser/compress/issue_973/this_binding_side_effects/output.js [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
     (0, foo.bar)();
     (0, eval)("console.log(foo);");
 })();
-(function (foo) {
-    var eval = console;
+(function(foo) {
+    var eval1 = console;
     foo();
     (0, foo.bar)();
-    (0, eval)("console.log(foo);");
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/terser/compress/issue_973/this_binding_side_effects/output.js:10]: This is also an indirect eval, so we can mangle names

**Review by kdy1 — COMMENTED** [review]
swc-bump:
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/tests/fixture/next/feedback-3/579-dcac359116b2707c/output.js [hunk 3]
```diff
@@ -0,0 +1,2415 @@
...
+                            t.exports = r(9720);
+                        }
+                    }, e = {};
+                    function o(r) {
+                        var n = e[r];
+                        if (void 0 !== n) return n.exports;
+                        var i = e[r] = {
+                            exports: {}
+                        };
+                        return t[r](i, i.exports, o), i.exports;
```
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/fixture/next/feedback-3/579-dcac359116b2707c/output.js:1253]: This is the fix

**Comment by Smrtnyk** [comment]
Does this also fix https://github.com/swc-project/swc/issues/5068?

**Comment by kdy1** [comment]
Actually this is a workaround and we should fix #5068 ideally

File: crates/swc_ecma_transforms_base/src/rename/eval.rs [hunk 4]
```diff
@@ -31,9 +31,17 @@ impl Visit for EvalFinder {
...
 
-    fn visit_ident(&mut self, i: &Ident) {
-        if i.sym == js_word!("eval") {
-            self.found = true;
+    fn visit_callee(&mut self, c: &Callee) {
+        c.visit_children_with(self);
+
+        if let Callee::Expr(e) = c {
+            if let Expr::Ident(Ident {
+                sym: js_word!("eval"),
```
Comment (jridgewell) [inline crates/swc_ecma_transforms_base/src/rename/eval.rs:39]: Not sure how you represent optional chains, but `eval?.()` is not a direct eval either.

File: crates/swc_ecma_minifier/tests/terser/compress/issue_976/eval_unused/output.mangleOnly.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
-function f1(a, eval, c, d, e) {
-    return a("c") + eval;
+function f1(n, eval, c, r, t) {
+    return n("c") + eval;
```
Comment (jridgewell) [inline crates/swc_ecma_minifier/tests/terser/compress/issue_976/eval_unused/output.mangleOnly.js:2]: Nit: this `eval` can be renamed.

**Review by jridgewell — APPROVED** [review]

File: crates/swc_ecma_transforms_base/src/rename/eval.rs (hunk 4 above)
Comment (kdy1) [reply crates/swc_ecma_transforms_base/src/rename/eval.rs:39]: image

Nice catch, thank you!

Comment (kdy1) [reply crates/swc_ecma_transforms_base/src/rename/eval.rs:39]: Oh it was `OptCall` which does not contain `Callee`

**Pull request merged**